### ISSUE TITLE
Support Multipart Forms

### ISSUE THREAD

**Issue by FallenValkyrie** [algesten/ureq#698]
Currently I cannot find any support for sending multipart Form data (`multipart/form-data`), It would be great If that was possible.

**Comment by johan-bjareholt** [comment]
It's quite easy to do youself, just:

1. generate a random boundary
2. Set HTTP content type header with said boundary like this "multipart/form-data; boundary={}"
3. For each part, loop this to create the HTTP body
    - write start delimiter --BOUNDARY
    - write "Content-Disposition" header and other optional HTTP headers for this specific part
    - write \r\n\r\n to end header section
    - write part body
    - write \r\n
 4. Write ending boundary --BOUNDARY--

I have done this, but it is unfortunately not open source and has multiple very specific details that only fits my rare use-case so it's not really suitable for upstreaming.

**Comment by algesten** [comment]
This needs doing for ureq 3.x as well. Let's study the API in reqwest (and others?). Maybe behind a feature flag depending on how messy it becomes.

Help wanted.

**Comment by algesten** [comment]
Yeah. That's neat. We should be able to reuse the `AsSendBody` trait for that.

**Comment by gmacon** [comment]
I ran into this today, and tonight took a stab at moving part of the implementation from the now-unmaintaned multipart crate (since it is compatibly licensed) into ureq. My work in progress is at https://github.com/gmacon/ureq/tree/multipart, but there's something wrong with the content-length calculations, and I don't feel like debugging it tonight.

**Comment by algesten** [comment]
Looks good. I skimmed the code, and two things popped out.

1. Cows 🐮  in public interfaces. This is a type I don't want to subject ureq's users to. Generally the philosophy of ureq is to make the API as accommodating to new user's of Rust as possible. From a high level view, I don't see what `Cow` give us here. It looks like the entire built multipart could have lifetimes and accept never being `'static`, but I might be missing something.
2. Use [`AsSendBody`](https://docs.rs/ureq/3.0.12/ureq/trait.AsSendBody.html) – we don't need a new function to send these bodies. Once constructed, it should fit in to the "general" body handling.

**Comment by algesten** [comment]
I'm also interested in looking at reqwest's API here, because it looks very neat in the examples.

**Comment by gmacon** [comment]
I can definitely look at those things; this started as an experiment to see if it was worth trying to use this existing code as a base. I think we do need a separate function, parallel to `send_form`, in order to set up the Content-Length and Content-Type headers, unless there's something I missed...

**Comment by gmacon** [comment]
OK, I have found and fixed the bug, so the example is now working!

I started thinking about how to make the API more like reqwest's. Here's my current thinking:

1. Rename `Multipart` to `Form`.
2. Change field names to `Into<Cow<'static, str>>`, matching reqwest. The benefit of using a `Cow` here is that it allows the `Form` to take ownership of a *temporary* String that would otherwise not live long enough. For example: `form.add_text(format!("..."), ...)` works with `Cow`, but using `&str` would require the caller to save the formatted result into a local variable. Though `Cow` appears in the API, the caller should never need to create one themselves; that will happen inside the function when `name.into()` is called. Changing to `'static` gets rid of the `'n` lifetime parameter.
3. Change to eager checking for existence of files. (This gets rid of `LazyError` and `IntoCowPath`.)
4. Change `SendBody` to track the size of the body in more cases (at least supporting `File` and `&File`).
5. Use `AsSendBody` as the trait bound on the data for `add_stream`.
6. Implement `AsSendBody` for the renamed `Form`. The implementation will parallel the current `prepare`, but the `PreparedFields` type (and subtypes) will become crate-private. After `<Form as AsSendBody>::as_body` returns, the `Form` will be *empty*.
7. I currently intend to keep `send_multipart_form` to set `Content-Type: multipart/form-data; boundary="..."` on the request, parallel to `send_form`, which sets `Content-Type: application/x-www-form-urlencoded`.

**Comment by algesten** [comment]
> Change SendBody to track the size of the body in more cases (at least supporting File and &File).

This is an interesting idea, that I'd like to discuss in a separate PR. The other day I was wondering whether `SendBody` should track both `content-length` and `content-type` where appropriate.

**Comment by algesten** [comment]
> * Rename `Multipart` to `Form`.

Check!

> * Change field names to `Into<Cow<'static, str>>`, matching reqwest. The benefit of using a `Cow` here is that it allows the `Form` to take ownership of a _temporary_ String that would otherwise not live long enough. For example: `form.add_text(format!("..."), ...)` works with `Cow`, but using `&str` would require the caller to save the formatted result into a local variable. Though `Cow` appears in the API, the caller should never need to create one themselves; that will happen inside the function when `name.into()` is called. Changing to `'static` gets rid of the `'n` lifetime parameter.

Ideally I want the user to be able to read the docs without having to understand `Cow`. I hear what you're saying about temporary `String`. What if we make the entire `Form` have the lifetime of its arguments?

> * Change to eager checking for existence of files. (This gets rid of `LazyError` and `IntoCowPath`.)

This makes sense to me. I wonder what the arguments is for having lazy errors. Wonder if reqwest documented that decision.

> * Change `SendBody` to track the size of the body in more cases (at least supporting `File` and `&File`).

See comment above.

> * Use `AsSendBody` as the trait bound on the data for `add_stream`.

Check!

> * Implement `AsSendBody` for the renamed `Form`. The implementation will parallel the current `prepare`, but the `PreparedFields` type (and subtypes) will become crate-private. After `<Form as AsSendBody>::as_body` returns, the `Form` will be _empty_.

Check!

> * I currently intend to keep `send_multipart_form` to set `Content-Type: multipart/form-data; boundary="..."` on the request, parallel to `send_form`, which sets `Content-Type: application/x-www-form-urlencoded`.

What if we make the multipartedness a property/switch on the `Form` itself?

`Form::builder().multipart().add_file(…).build()`

Maybe it would make sense to go even further: `Form<Plain>` and `Form<Multipart>` that exposes a different set of API depending on mode? Like, you wouldn't really attach a file to a plain form (name tbd).

**Comment by gmacon** [comment]
> > * Change field names to `Into<Cow<'static, str>>`, matching reqwest. The benefit of using a `Cow` here is that it allows the `Form` to take ownership of a _temporary_ String that would otherwise not live long enough. For example: `form.add_text(format!("..."), ...)` works with `Cow`, but using `&str` would require the caller to save the formatted result into a local variable. Though `Cow` appears in the API, the caller should never need to create one themselves; that will happen inside the function when `name.into()` is called. Changing to `'static` gets rid of the `'n` lifetime parameter.
> 
> Ideally I want the user to be able to read the docs without having to understand `Cow`. I hear what you're saying about temporary `String`. What if we make the entire `Form` have the lifetime of its arguments?

I feel like we're talking past each other... maybe an example would be helpful? What do you think about https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=bdbe273774755ece3fd8da3d6eb3f2ac? That suggests to me that if we don't want `Cow`, we should switch to `String`. That would require some unnecessary copies, but the field names should be small and it wouldn't be that expensive.

> > * Change to eager checking for existence of files. (This gets rid of `LazyError` and `IntoCowPath`.)
> 
> This makes sense to me. I wonder what the arguments is for having lazy errors. Wonder if reqwest documented that decision.

This wasn't something reqwest did, this was something that multipart did. Programming with errors is a clever idea, and I appreciate it aesthetically, but I don't think it's worth the complexity in this case.

> > * I currently intend to keep `send_multipart_form` to set `Content-Type: multipart/form-data; boundary="..."` on the request, parallel to `send_form`, which sets `Content-Type: application/x-www-form-urlencoded`.
> 
> What if we make the multipartedness a property/switch on the `Form` itself?
> 
> `Form::builder().multipart().add_file(…).build()`
> 
> Maybe it would make sense to go even further: `Form<Plain>` and `Form<Multipart>` that exposes a different set of API depending on mode? Like, you wouldn't really attach a file to a plain form (name tbd).

I'm not sure there's any harm in using a multipart/form-data for a form without a file attachment, but it's not what a web browser would do. If we adopt the suggestion in #1098 of capturing the content-type, that would solve the need for an additional function here.

**Comment by algesten** [comment]
`content-length` capturing is now in main via #1100

**Comment by algesten** [comment]
I took the liberty to open your starting code as a PR #1101  – I want to review the API.

**Comment by algesten** [comment]
@gmacon your original code in #1101, I think we can close that PR. I made a stab at the API I think we want in #1102 

Left to do is implement reading the Form body in a streaming way with a minimum of allocations.
